Adding docs for filtered include feature and updating query samples (#2260)
dotnet/EntityFramework.Docs: entity-framework/core/querying/related-data.md

> [!CAUTION]
> Since version 3.0.0, each `Include` will cause an additional JOIN to be added to SQL queries produced by relational providers, whereas previous versions generated additional SQL queries. This can significantly change the performance of your queries, for better or worse. In particular, LINQ queries with an exceedingly high number of `Include` operators may need to be broken down into multiple separate LINQ queries in order to avoid the cartesian explosion problem.

### Filtered include

> [!NOTE]

> > This feature will be introduced in EF Core 5.0.

Reviewer: Just write is....

Reviewer: Also it is already introduced. It is that 5.0 is not released yet. so will be would be incorrect. :trollface:

Reviewer 2: I am very tempted to launch into this discussion! But I am strong.

> The filtered include feature enables applying operations directly on the included collection navigation, which allows for filtering and sorting of the results.

Reviewer: enumerable operations.

> The filtered include feature enables applying operations directly on the included collection navigation, which allows for filtering and sorting of the results.

Reviewer: ```suggestion
When applying Include to load related data, you can apply certain enumerable operations on the included collection navigation, which allows for filtering and sorting of the results.
```

Supported operations are: `Where`, `OrderBy`, `OrderByDescending`, `ThenBy`, `ThenByDescending`, `Skip`, and `Take`.

> Those operations should be applied inside the Include method:

Reviewer: applied on the collection navigation in the lambda passed to include method.

> Those operations should be applied inside the Include method:

Reviewer: ```suggestion
Such operations should be applied on the collection navigation in the lambda passed to the Include method. As shown in example below:
```

[!code-csharp[Main](../../../samples/core/Querying/RelatedData/Sample.cs#FilteredInclude)]

> Each included navigation allows only one unique set of filter operations. In cases where multiple Include operations are applied on a given collection navigation (`blog.Posts` in the examples below), filter operations can only be applied to one of them: 

Reviewer: ```suggestion
Each included navigation allows only one unique set of filter operations. In cases where multiple Include operations are applied for a given collection navigation (`blog.Posts` in the examples below), filter operations can be specified on one of them:
```

Reviewer: We should just merge both of these in one paragraph. Also it is not like all or none. you can specifying the filter on any occurrence of collection navigation as long as it is same.

Reviewer: I read your comment above, if you think lower priority and we should keep separate then it is fine. Though consider above suggestion.

  context.People.Include("School").ToList()
  ```

> ### Filtered include

Reviewer 2: We should consider putting this just below "Including multiple levels", above "Include on derived types". It connects more naturally to including multiple levels and seems generally more relevant to people.

> [!NOTE]

> > This feature was introduced in EF Core 5.0.

Reviewer 2: Not sure what we do for upcoming releases, but maybe "will be introduced"? Mention the specific preview version as well?

Author: Will add preview version once it actually ships ;)

Reviewer: I think it is fine to write introduced in 5.0. Just additional maintenance cost otherwise. The feature will also end up on https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-5.0/whatsnew to say which exact preview it was released in.

> Filtered include feature enables applying some operations directly on the included collection navigation, which allows for filtering and sorting of the results.

Reviewer 2: ```suggestion
The filtered include feature enables applying operations directly on the included collection navigation, which allows for filtering and sorting of the results.
```

Supported operations are: `Where`, `OrderBy`, `OrderByDescending`, `ThenBy`, `ThenByDescending`, `Skip`, and `Take`.

Those operations should be applied inside the Include method:

[!code-csharp[Main](../../../samples/core/Querying/RelatedData/Sample.cs#FilteredInclude)]

> Each included navigation allows only one unique set of filter operations. In case when Include operation is applied more than once on a given collection navigation (`blog.Posts` in the examples below), filter operations should only be applied on one of them. 

Reviewer 2: ```suggestion
Each included navigation allows only one unique set of filter operations. In cases where multiple Include operations are applied on a given collection navigation (`blog.Posts` in the examples below), filter operations must only be applied to one of them. 
```

Reviewer 2: BTW what happens if I accidentally put non-identical operations, do we throw informatively?

Reviewer: Non-identical filter we throw.
It should be can rather than should/must. If you use same collection multiple time and specify same filter then it works.

Reviewer 2: Yeah, there's a sentence saying that below. Maybe it should be merged, I agree it's not totally clear as-is.

Author: we throw a decent error:
Different filters: '{filter1}' and '{filter2}' have been applied on the same included navigation. Only one unique filter per navigation is allowed. For more information on including related data, see http://go.microsoft.com/fwlink/?LinkID=746393.

Author: I anticipate main use will be to apply it only once and use regular include for all other instances. The case when u apply the same filter multiple times is probably only useful in the code generation scenarios or maybe copy-paste, brain-dead coding. I'm happy to merge the paragraphs together, but i think the second scenario is definitely lower priority

[!code-csharp[Main](../../../samples/core/Querying/RelatedData/Sample.cs#MultipleLeafIncludesFiltered1)]

> Alternatively, identical operations can be applied for each navigation that is included multiple times.

Reviewer 2: ```suggestion
Alternatively, identical operations can be applied for each navigation that is included multiple times:
```

[!code-csharp[Main](../../../samples/core/Querying/RelatedData/Sample.cs#MultipleLeafIncludesFiltered2)]

> 

Author: many of the filtered include scenarios translate to apply, so sqlite will throw - should we add a warning/caution for that here?

Reviewer: for filtered includes, take works in sqlite right?

Author: Nope - 

```
SELECT "l"."Id", "l"."Date", "l"."Name", "l"."OneToMany_Optional_Self_Inverse1Id", "l"."OneToMany_Required_Self_Inverse1Id", "l"."OneToOne_Optional_Self1Id", "t"."Id", "t"."Date", "t"."Level1_Optional_Id", "t"."Level1_Required_Id", "t"."Name", "t"."OneToMany_Optional_Inverse2Id", "t"."OneToMany_Optional_Self_Inverse2Id", "t"."OneToMany_Required_Inverse2Id", "t"."OneToMany_Required_Self_Inverse2Id", "t"."OneToOne_Optional_PK_Inverse2Id", "t"."OneToOne_Optional_Self2Id"
FROM "LevelOne" AS "l"
OUTER APPLY (
    SELECT "l0"."Id", "l0"."Date", "l0"."Level1_Optional_Id", "l0"."Level1_Required_Id", "l0"."Name", "l0"."OneToMany_Optional_Inverse2Id", "l0"."OneToMany_Optional_Self_Inverse2Id", "l0"."OneToMany_Required_Inverse2Id", "l0"."OneToMany_Required_Self_Inverse2Id", "l0"."OneToOne_Optional_PK_Inverse2Id", "l0"."OneToOne_Optional_Self2Id"
    FROM "LevelTwo" AS "l0"
    WHERE "l"."Id" = "l0"."OneToMany_Optional_Inverse2Id"
    ORDER BY "l0"."Name"
    LIMIT 3
) AS "t"
ORDER BY "l"."Id", "t"."Name", "t"."Id"
```

Where works, OrderBy.AsQueryable works, Skip/Take fails due to APPLY, naked OrderBy fails due to a bug

Reviewer: Something is wrong then. We added whole support for RowNumberExpression to convert apply to join when there is only take. https://github.com/dotnet/efcore/pull/17293
There is also test which converts Take(4) in select many. At the minimum query you posted above has nothing in it which stops it from getting converted to join.

Also related to it: If we fix https://github.com/dotnet/efcore/issues/17326, then skip would also convert to number and most of it should get translated to Sqlite. (For filtered include case, everything would work).

Author: will investigate what's up

Author: ```
ctx.LevelOne.SelectMany(l1 => l1.OneToMany_Optional1.OrderBy(x => x.Name).Take(3).ToList()).ToList();
```

still translates to apply on sqlite (and fails).

Reviewer: File an issue and don't add any warning here.
